zebra-network/src/peer_set/initialize.rs

```diff
@@ -868,18 +868,18 @@ where
                     "opening an outbound peer connection"
                 );
 
+                // Try to get the next available peer for a handshake.
+                //
+                // candidates.next() has a short timeout, and briefly holds the address
+                // book lock, so it shouldn't hang.
+                //
+                // Hold the lock for as short a time as possible.
+                let candidate = { candidates.lock().await.next().await };
```

Comment:
I think this specific fix might revert a previously fixed deadlock or performance issue.

The crawler needs to run independently of candidate set locks and rate-limits, so that it can get new peer addresses to add to the address book. But this change blocks the crawler loop on the candidate set rate-limit.

Instead, I would suggest documenting the existing calling code and called method using a security comment.